segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/InMemoryStorage.java

```diff
@@ -195,6 +196,11 @@ public boolean supportsTruncation() {
         return false;
     }
 
+    @Override
+    public Iterator<SegmentProperties> listSegments() {
+        return this.streamSegments.values().stream().map( s -> s.getInfo()).iterator();
```

Comment:
Did you notice that `streamSegments` is guarded by a lock? 

This tells you two things:
1. Wrap this call in a `synchronized` block
2. Make a copy of the `values()` collection in that block and iterate over the copy. That will prevent your iterator from failing if some other thread modifies the collection while you iterate through it.